### PR TITLE
GUACAMOLE-611: Selectively fall through to other extensions when authentication fails

### PR DESCRIPTION
**Pull request by mike-jumper**
This change adds a new property, `skip-if-unavailable`, which allows specific authentication providers to be skipped in the event of an internal error. In situations where a Guacamole deployment involves multiple authentication providers, administrators can define this property to explicitly allow backend failures  without aborting authentication entirely.

For example, if a deployment is configured to use both LDAP and MySQL, but the network to the MySQL server is known to occasionally go down, the administrator might specify:

```
skip-if-unavailable: mysql
```

to allow LDAP to continue to be used even if connection errors prevent authentication using MySQL. Specifying multiple auth providers is also possible using commas:

```
skip-if-unavailable: mysql, ldap
```

If this property is specified, and an authentication provider is skipped due to an internal error, a warning is logged advising the administrator that this has occurred:

> The "postgresql" authentication provider has been skipped due to an internal error. If this is unexpected or you are the developer of this authentication provider, you may wish to enable debug-level logging: ...

If this property is *not* specified and authentication is halted entirely due to an internal error, a warning is logged advising the administrator that they may wish to set `skip-if-unavailable` if such failures are expected:

> The "postgresql" authentication provider has encountered an internal error which will halt the authentication process. If this is unexpected or you are the developer of this authentication provider, you may wish to enable debug-level logging. If this is expected and you wish to ignore such failures in the future, please set "skip-if-unavailable: postgresql" within your guacamole.properties.

### CODE REVIEW

File: guacamole/src/main/java/org/apache/guacamole/extension/AuthenticationProviderFacade.java [hunk 1]
```diff
@@ -48,6 +48,16 @@
...
 
+    /**
+     * The set of identifiers of all authentication providers whose internal
+     * failures should be tolerated during the authentication process. If the
+     * identifier of this authentication provider is within this set, errors
+     * during authentication will result in the authentication provider being
+     * ignored for that authentication attempt. By default, errors during
+     * authentication halt the authentication process entirely.
+     */
+    private final Set<String> tolerateFailures;
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply guacamole/src/main/java/org/apache/guacamole/extension/AuthenticationProviderFacade.java]: Yep. The identifier of the underlying `AuthenticationProvider` will not be available until after construction (as it's provided by the implementation of `getIdentifier()`), and then only if the underlying `AuthenticationProvider` is successfully instantiated.

File: guacamole/src/main/java/org/apache/guacamole/extension/AuthenticationProviderFacade.java [hunk 2]
```diff
@@ -97,18 +119,122 @@ public Object getResource() throws GuacamoleException {
...
+        String identifier = getIdentifier();
+        logger.warn("The \"{}\" authentication provider has encountered an "
+                + "internal error which will halt the authentication "
+                + "process. If this is unexpected or you are the developer of "
+                + "this authentication provider, you may wish to enable "
+                + "debug-level logging. If this is expected and you wish to "
+                + "ignore such failures in the future, please set \"{}: {}\" "
+                + "within your guacamole.properties.",
+                identifier, ExtensionModule.SKIP_IF_UNAVAILABLE.getName(),
+                identifier);
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply guacamole/src/main/java/org/apache/guacamole/extension/AuthenticationProviderFacade.java]: I believe the exception handling within the webapp will automatically log the details at the debug level.

File: guacamole/src/main/java/org/apache/guacamole/extension/AuthenticationProviderFacade.java [hunk 3]
```diff
@@ -97,18 +119,122 @@ public Object getResource() throws GuacamoleException {
...
+    }
+
     @Override
     public AuthenticatedUser authenticateUser(Credentials credentials)
             throws GuacamoleException {
 
         // Ignore auth attempts if no auth provider could be loaded
         if (authProvider == null) {
             logger.warn("Authentication attempt denied because the authentication system could not be loaded. Please check for errors earlier in the logs.");
```
[the remark this answers is not in the document]
Comment (necouchman) [reply guacamole/src/main/java/org/apache/guacamole/extension/AuthenticationProviderFacade.java]: > "denied" is no longer accurate. The comment above the if block is more correctly worded, stating that the auth attempt is being ignored. Rewording the warning may be in order.

Sounds good - I'll re-review, but I think once that rewording is done it'll be ready for merge.

[the remark this answers is not in the document]
Comment (mike-jumper) [reply guacamole/src/main/java/org/apache/guacamole/extension/AuthenticationProviderFacade.java]: OK - I've updated the warnings.

**Pull request merged**